docs/_posts/2017-11-30-nullannotations.md

```diff
@@ -0,0 +1,65 @@
+
+We have already started to annotate our core packages so binding developers can benefit from them.
+
+### Real life examples
+
+One simple example is the [ThingHandler:handleCommand(ChannelUID channelUID, Command command)](https://github.com/eclipse/smarthome/blob/master/bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/ThingHandler.java#L99) which guarantees to binding developers that both arguments are **not** `null` and so they can simply implement their logic by going through the IDs via `switch (channelUID.getId())` without running into a NPE.
+
+{:.center}
+![NonNull annotation]({{ "/img/blog/2017-11-30-nullannotations_handleCommand.png" | absolute_url }})
+
+For bindings which use bridge things and call [BaseThingHandler:getBridge()](https://github.com/eclipse/smarthome/blob/master/bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/BaseThingHandler.java#L584) we now have the `@Nullable` annotation in place that reminds developers to do a proper `null` check before using the return value of this method.
+If you omit the `null` check, developers will get a warning as shown below.
+
+{:.center}
+![Nullable annotation with warning]({{ "/img/blog/2017-11-30-nullannotations_bridgeNullable.png" | absolute_url }})
+
+After the `null` check has been implemented the warning is gone:
+
+{:.center}
+![Nullable annotation without warning]({{ "/img/blog/2017-11-30-nullannotations_bridgeNullableCorrected.png" | absolute_url }})
+
+If developers violate the `null` specifications, the IDE will immediately show an error.
+In the example below the `BaseThingHandler:updateState(String channelID, State state)` method requires both parameters to be unequal to `null`.
+But since in the `else` branch the variable used as a second parameter in this method call is set to `null`, the IDE infers that `null` is passed as an argument to the method and shows an error.
+
+{:.center}
+![Null violation]({{ "/img/blog/2017-11-30-nullannotations_error.png" | absolute_url }})
+
+If you are interested to see a fully annotated binding, please have a look at the [Philips Hue binding](https://github.com/eclipse/smarthome/tree/master/extensions/binding/org.eclipse.smarthome.binding.hue).
+
+
+### Outlook: Third party code
+
+If we make use of external libraries and use their return values in our code the compiler cannot decide whether these values are always unequal to `null` or might become `null` at some point.
+This will generate a warning that is currently switched off by our setup.
+In the future we are planing to support *external annotations* which are files that contain information about the method signatures of external libraries.
+The [Last NPE project](http://last-npe.org) has started a crowd-sourcing initative to create these external annotations and we are planing to contribute our work to this project.
```

Comment:
Err, that's nice of you... But isn't it rather the plan to _use_ this project's deliverables?